### PR TITLE
Improve fp16 performance of GPT-2's logits MatMul while using BeamSearch

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
While using BeamSearch, allow for padding the logits MatMul's weight such that it becomes a multiple of 8 (usally the vocab size isn't necessarily a multiple of 8) and thus allowing CuBlas to leverage tensor cores (for the fp16 case) and this should speed-up one run of the decoder subgraph by several micro-seconds (these savings will be accumulated across different decoding runs for a language model). 

Performance evaluation:
The logits MatMul is faster by several micro-seconds (confirmed by profiling)

This is currently limited to GPT2 Beam Search alone. In future, we may easily extend this capability to different model architectures/decoding strategies as needed.

### Motivation and Context
Make fp16 GPT-2 BeamSearch faster

### CODE REVIEW

**Review by wangyems — COMMENTED** [review]
lgtm

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 1]
```diff
@@ -407,6 +418,65 @@ def shape_inference(onnx_path: str, use_external_data_format: bool = True):
...
+    if matmul_node.op_type != "MatMul":
+        return False
+    
+    # The logits MatMul weight MUST be an initializer
+    logits_weight = decoder_model.get_initializer(matmul_node.input[1])
+    if logits_weight is None:
+        return False
+
+    # The logits MatMul weight MUST be fp16
+    if (logits_weight.data_type != TensorProto.DataType.FLOAT16):
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:448]: Could we also handle a case of initializer --> Transpose --> MatMul. It is because the initializer is same as the one in embedding layer. So using Transpose could reduce model size.

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 2]
```diff
@@ -407,6 +418,65 @@ def shape_inference(onnx_path: str, use_external_data_format: bool = True):
...
+    assert logits_output_name in output_name_to_node
+
+    matmul_node = output_name_to_node[logits_output_name]
+    # Sanity check - the logits need to be produced by a MatMul node
+    if matmul_node.op_type != "MatMul":
+        return False
+    
+    # The logits MatMul weight MUST be an initializer
+    logits_weight = decoder_model.get_initializer(matmul_node.input[1])
+    if logits_weight is None:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:444]: Could we also handle a case of initializer --> Transpose --> MatMul. It is because the initializer is same as the one in embedding layer. So using Transpose could reduce model size.

File: docs/ContribOperators.md [hunk 3]
```diff
@@ -404,6 +404,8 @@ This version of the operator has been available since version 1 of the 'com.micr
 <dd>no repeat ngrams size</dd>
 <dt><tt>pad_token_id</tt> : int (required)</dt>
 <dd>The id of the padding token</dd>
+<dt><tt>vocab_size</tt> : int</dt>
+<dd>Size of the vocabulary.If not provided, it will be inferred from the decoder subgraph's output shape</dd>
```
Comment (tianleiwu) [inline docs/ContribOperators.md:408]: nit: add a space before If.

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 4]
```diff
@@ -171,6 +173,15 @@ def parse_arguments(argv: Optional[List[str]] = None) -> argparse.Namespace:
     )
     output_group.set_defaults(run_shape_inference=False)
 
+    output_group.add_argument(
+        "-plw", 
+        "--pad_logits_matmul_weight",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:178]: --pad_vocab_size might be better

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 5]
```diff
@@ -407,6 +418,65 @@ def shape_inference(onnx_path: str, use_external_data_format: bool = True):
...
+    # The logits MatMul weight MUST be 2-dimensional
+    if (len(logits_weight.dims) != 2):
+        return False
+
+    # Pad and over-write the initializer (if needed)
+    # TODO(hasesh): Can/Will the weights be transposed in any scenario ?
+    actual_vocab_size = logits_weight.dims[1]
+
+    if ((actual_vocab_size % 8) == 0):
+        return False
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:460]: How about return True here. Otherwise, there is warning in line 899.

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc [hunk 6]
```diff
@@ -230,32 +230,36 @@ Status ProcessLogits(const OrtValue& logits,                                 //
...
   gsl::span<T>& next_token_logits = beam_state->next_token_logits;
-
   // TODO(tianleiwu): use one kernel to replace a loop of memory copy.
   if (input_length > 1 || logits_batch_size == batch_size) {
-    const CudaT* current_logits = logits_data + (input_length - 1) * vocab_size;
+    const CudaT* current_logits = logits_data + (input_length - 1) * padded_vocab_size;
     for (int i = 0; i < batch_beam_size; i++) {
       gsl::span<const T> source(reinterpret_cast<const T*>(current_logits), vocab_size);
       gsl::span<T> target = next_token_logits.subspan(static_cast<size_t>(i) * vocab_size, vocab_size);
       CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target.data(), source.data(), sizeof(T) * vocab_size,
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:257]: Here only vocab_size is copied. Why not copy padded_vocab_size?

File: docs/ContribOperators.md (hunk 3 above)
Comment (tianleiwu) [inline docs/ContribOperators.md:408]: Do we need also update the shape inference for BeamSearch? One is for the operator def, and another is in symbolic shape inference.

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/python/tools/transformers/convert_generation.py:444]: Handled

**Comment by tianleiwu** [comment]
A few pipeline failed: Use --gen_doc to update document; and use [black](https://pypi.org/project/black/) to format python code.

**Comment by hariharans29** [comment]
> A few pipeline failed: Use --gen_doc to update document; and use [black](https://pypi.org/project/black/) to format python code.

Yeah, I am aware of this. I working on this. WIll mark the PR WIP for now.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 7]
```diff
@@ -451,11 +451,20 @@ void BeamSearchShapeInference(ONNX_NAMESPACE::InferenceContext& ctx) {
...
+      int64_t vocab_size = vocab_size_attr ? static_cast<int64_t>(vocab_size_attr->i()) : -1;
+
       ONNX_NAMESPACE::TensorShapeProto scores_shape;
       scores_shape.add_dim()->set_dim_value(max_length_value - sequence_length);
       scores_shape.add_dim()->set_dim_value(batch_size);
       scores_shape.add_dim()->set_dim_value(num_beams_value);
-      scores_shape.add_dim();  // vocab_size is unknown
+      if (vocab_size != -1) {
+        // vocab_size is provided by the user - use that
+        scores_shape.add_dim()->set_dim_value(vocab_size);
```
Comment (hariharans29) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:463]: Made BeamSearch's shape inference handle the case where vocab_size is provided - in this case, we can do complete shape inference (and not let vocab_size be unknown). If there are other places that need similar updates, please let me know (like SymbolicShapeInference ?)

**Review by tianleiwu — APPROVED** [review]
